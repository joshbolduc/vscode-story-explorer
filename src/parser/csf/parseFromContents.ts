import { parse } from '@babel/parser';
import traverse from '@babel/traverse';
import type { Binding, NodePath } from '@babel/traverse';
import {
  ClassDeclaration,
  Expression,
  FunctionDeclaration,
  Identifier,
  isExportSpecifier,
  isIdentifier,
  isTSSatisfiesExpression,
  ObjectExpression,
  SourceLocation,
  TSDeclareFunction,
  VariableDeclarator,
} from '@babel/types';
import type { Location } from '../../types/Location';
import { sourceLocationToLocation } from '../sourceLocationToLocation';

interface RawMeta {
  id?: string;
  location: Location | undefined;
  properties: Record<string, unknown>;
  declared: boolean;
}

export interface RawStory {
  exportName: string;
  location: Location | undefined;
  properties: Record<string, unknown>;
}

interface RawResult {
  meta: RawMeta;
  stories: RawStory[];
}

const parseMetaObjectExpression = (
  objectExpression: NodePath<ObjectExpression>,
) => {
  return objectExpression
    .get('properties')
    .reduce<Record<string, unknown>>((acc, property) => {
      if (property.isObjectProperty()) {
        const key = property.node.key;
        const name = 'name' in key ? key.name : undefined;

        if (name) {
          const evalResult = property.get('value').evaluate();
          if (evalResult.confident) {
            acc[name] = evalResult.value;
          }
        }
      }
      return acc;
    }, {});
};

const createStory = (
  exportName: string,
  localName: string,
  location: SourceLocation | null | undefined,
  path: NodePath,
) => {
  const newStory: RawStory = {
    exportName,
    location: location
      ? {
          start: location.start,
          end: location.end,
        }
      : undefined,
    properties: {},
  };

  // find references
  const bindings = path.scope.getAllBindings() as Record<string, Binding>;
  const referencePaths = bindings[localName]?.referencePaths;
  referencePaths?.forEach((referencePath) => {
    if (
      !referencePath.isIdentifier() ||
      !referencePath.parentPath.isMemberExpression()
    ) {
      return;
    }

    const memberExpression = referencePath.parentPath;

    const memberProperty = memberExpression.node.property;

    if (!isIdentifier(memberProperty)) {
      return;
    }

    const propertyName = memberProperty.name;

    const assignmentExpression = memberExpression.parentPath;
    if (!assignmentExpression.isAssignmentExpression()) {
      return;
    }

    const right = assignmentExpression.get('right');

    const evalResult = right.evaluate();

    if (evalResult.confident) {
      newStory.properties[propertyName] = evalResult.value;
    }
  });

  return newStory;
};

const getNestedExpression = (
  node: NodePath<
    | ClassDeclaration
    | FunctionDeclaration
    | TSDeclareFunction
    | Expression
    | null
    | undefined
  >,
) => {
  let expr = node;

  while (expr.isTSAsExpression() || isTSSatisfiesExpression(expr.node)) {
    expr = expr.get('expression') as NodePath<Expression>;
  }

  return expr;
};

export const parseFromContents = (contents: string): RawResult => {
  const parsedFile = parse(contents, {
    sourceType: 'unambiguous',
    plugins: ['jsx', 'typescript'],
    errorRecovery: true,
  });

  const meta: RawMeta = {
    location: undefined,
    properties: {},
    declared: false,
  };
  const stories: RawStory[] = [];

  const addMetaObjectProperties = (
    properties: Record<string, unknown>,
    location: Location | undefined,
  ) => {
    meta.declared = true;
    Object.assign(meta.properties, properties);

    meta.location = location;
    if (typeof properties.id === 'string') {
      meta.id = properties.id;
    }
  };

  const addFromBinding = (expression: NodePath<Identifier>) => {
    const binding = expression.scope.getBinding(expression.node.name);
    const bindingPath = binding?.path;
    if (bindingPath?.isVariableDeclarator()) {
      const init = bindingPath.get('init');

      const nestedExpression = getNestedExpression(init);

      if (nestedExpression.isObjectExpression()) {
        const sourceLoc = bindingPath.node.loc;
        addMetaObjectProperties(
          parseMetaObjectExpression(nestedExpression),
          sourceLocationToLocation(sourceLoc),
        );
      }
    }
  };

  const createStoryFromDeclarationPath = (
    path: NodePath<FunctionDeclaration | VariableDeclarator>,
  ) => {
    const id = path.node.id;

    if (isIdentifier(id)) {
      const exportName = id.name;
      const location = id.loc;

      const newStory = createStory(exportName, exportName, location, path);
      stories.push(newStory);
    }
  };

  traverse(parsedFile, {
    ExportDefaultDeclaration: (path) => {
      const declaration = path.get('declaration');
      if (declaration.isObjectExpression()) {
        const sourceLoc = declaration.node.loc;
        addMetaObjectProperties(
          parseMetaObjectExpression(declaration),
          sourceLocationToLocation(sourceLoc),
        );
      } else if (declaration.isIdentifier()) {
        addFromBinding(declaration);
      } else if (
        declaration.isTSAsExpression() ||
        isTSSatisfiesExpression(declaration.node)
      ) {
        const expression = getNestedExpression(declaration);
        if (expression.isObjectExpression()) {
          const sourceLoc = declaration.node.loc;
          addMetaObjectProperties(
            parseMetaObjectExpression(expression),
            sourceLocationToLocation(sourceLoc),
          );
        } else if (expression.isIdentifier()) {
          addFromBinding(expression);
        }
      }
    },
    ExportNamedDeclaration: (path) => {
      path.node.specifiers.forEach((specifierNode) => {
        if (
          isExportSpecifier(specifierNode) &&
          isIdentifier(specifierNode.local) &&
          isIdentifier(specifierNode.exported)
        ) {
          const localIdentifier = specifierNode.local;
          const exportedIdentifier = specifierNode.exported;

          const bindings = path.scope.getAllBindings() as Record<
            string,
            Binding
          >;
          const binding = bindings[localIdentifier.name];
          const location = binding?.identifier.loc;
          const newStory = createStory(
            exportedIdentifier.name,
            localIdentifier.name,
            location,
            path,
          );
          stories.push(newStory);
        }
      });

      const declarationPath = path.get('declaration');
      if (declarationPath) {
        if (declarationPath.isFunctionDeclaration()) {
          createStoryFromDeclarationPath(declarationPath);
        } else if (declarationPath.isVariableDeclaration()) {
          declarationPath
            .get('declarations')
            .forEach((variableDeclarationPath) => {
              createStoryFromDeclarationPath(variableDeclarationPath);
            });
        }
      }
    },
  });

  return { meta, stories };
};
