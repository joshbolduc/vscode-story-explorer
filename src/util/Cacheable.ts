/**
 * An object that lazily performs a predefined operation and caches its result.
 */
export class Cacheable<T> {
  private cachedResult?: { result: T } | undefined;

  /**
   * Initialize a new Cacheable with a given callback.
   *
   * @param callback A callback that will be lazily invoked when the result is
   * requested, and whose return value will be cached and returned for
   * subsequent requests.
   */
  public constructor(private callback: () => T) {}

  /**
   * Get the result of the operation, invoking it if necessary. If the operation
   * has already been invoked, its result is returned.
   *
   * @returns The result of the operation, cached or otherwise.
   */
  public getResult() {
    if (!this.cachedResult) {
      this.cachedResult = { result: this.callback() };
    }

    return this.cachedResult.result;
  }

  /**
   * Get the cached result of the operation, if any, or a default value if the
   * result has not already been determined.
   */
  public getCachedResult(): T | undefined;
  public getCachedResult<U>(defaultValue: U): T | U;
  public getCachedResult<U>(defaultValue?: U) {
    return this.cachedResult ? this.cachedResult.result : defaultValue;
  }

  /**
   * Replace the callback used to computed the cached result. The cached result,
   * if any, is cleared.
   *
   * @param callback A callback that will be lazily invoked when the result is
   * requested, and whose return value will be cached and returned for
   * subsequent requests.
   */
  public setCallback(callback: () => T) {
    this.clear();
    this.callback = callback;
  }

  /**
   * Clears the cached result. The next time a result is requested, the
   * operation will be reinvoked and its return value will be cached again.
   */
  public clear() {
    this.cachedResult = undefined;
  }
}
