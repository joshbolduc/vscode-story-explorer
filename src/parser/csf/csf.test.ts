import { readFileSync } from 'fs';
import { join } from 'path';
import { sync } from 'fast-glob';
import { describe, expect, it } from 'vitest';
import { testBaseDir } from '../../../test/util/testBaseDir';
import { parseFromContents } from './parseFromContents';

describe('csf', () => {
  const tests = sync(['project/v6/src/**/*.stories.{js,jsx,ts,tsx}'], {
    cwd: testBaseDir,
  });

  tests.forEach((match) => {
    it(`parses story file ${match}`, () => {
      const storiesPath = join(testBaseDir, match);
      const contents = readFileSync(storiesPath).toString();
      const result = parseFromContents(contents);

      expect(result).toMatchSnapshot();
    });
  });
});
