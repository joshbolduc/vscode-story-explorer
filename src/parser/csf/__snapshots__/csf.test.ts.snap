// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`csf > parses story file project/v6/src/autoTitle/csf/EmptyDefaultExport.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 5, Col 17],
    "properties": {},
  },
  "stories": [
    {
      "exportName": "Primary",
      "location": [Ln 9, Col 13] - [Ln 9, Col 20],
      "properties": {
        "args": {
          "children": "Empty default export",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v6/src/autoTitle/csf/EmptyTitle.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "properties": {
      "title": "",
    },
  },
  "stories": [
    {
      "exportName": "Primary",
      "location": [Ln 11, Col 13] - [Ln 11, Col 20],
      "properties": {
        "args": {
          "children": "Story with empty title and prefix",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v6/src/autoTitle/csf/NoDefaultExport.stories.js 1`] = `
{
  "meta": {
    "declared": false,
    "location": undefined,
    "properties": {},
  },
  "stories": [
    {
      "exportName": "Primary",
      "location": [Ln 7, Col 13] - [Ln 7, Col 20],
      "properties": {
        "args": {
          "children": "No default export",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v6/src/stories/v6-only/TitleInterpolation.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 7, Col 15] - [Ln 10, Col 1],
    "properties": {
      "title": "Example/Title Interpolation",
    },
  },
  "stories": [
    {
      "exportName": "Primary",
      "location": [Ln 14, Col 13] - [Ln 14, Col 20],
      "properties": {
        "args": {
          "children": "Interpolation/Sample",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v6/src/stories/v6-only/TopHoisted.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 8, Col 1],
    "properties": {
      "title": "Top Hoisted",
    },
  },
  "stories": [
    {
      "exportName": "TopHoisted",
      "location": [Ln 12, Col 13] - [Ln 12, Col 23],
      "properties": {
        "args": {
          "children": "Top level hoisted story",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v6/src/stories/v6-only/TypeScriptIdentifierDefaultAsMeta.stories.ts 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 6, Col 6] - [Ln 8, Col 1],
    "properties": {
      "title": "Test/TypeScript/Identifier Default Exported as Meta",
    },
  },
  "stories": [
    {
      "exportName": "TypeScriptStory",
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v6/src/stories/v6-only/TypeScriptIdentifierDefaultSatisfiesAsMeta.stories.ts 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 6, Col 6] - [Ln 8, Col 1],
    "properties": {
      "title": "Test/TypeScript/Identifier exported satisfies as Meta",
    },
  },
  "stories": [
    {
      "exportName": "TypeScriptStory",
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v6/src/stories/v6-only/TypeScriptIdentifierDefaultSatisfiesMeta.stories.ts 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 6, Col 6] - [Ln 8, Col 1],
    "properties": {
      "title": "Test/TypeScript/Identifier exported satisfies Meta",
    },
  },
  "stories": [
    {
      "exportName": "TypeScriptStory",
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v6/src/stories/v6-only/TypeScriptIdentifierSatisfiesAsMeta.stories.ts 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 6, Col 6] - [Ln 8, Col 24],
    "properties": {
      "title": "Test/TypeScript/Identifier Satisfies as Meta",
    },
  },
  "stories": [
    {
      "exportName": "TypeScriptStory",
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/autoTitle/csf/ExtraTitle.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "properties": {
      "title": "Explicit title",
    },
  },
  "stories": [
    {
      "exportName": "Primary",
      "location": [Ln 11, Col 13] - [Ln 11, Col 20],
      "properties": {
        "args": {
          "children": "Story with explicit title and prefix",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/autoTitle/csf/nested/WithoutTitle.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "properties": {},
  },
  "stories": [
    {
      "exportName": "Primary",
      "location": [Ln 11, Col 13] - [Ln 11, Col 20],
      "properties": {
        "args": {
          "children": "Without Title/Sample",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/autoTitle/csf/nested/nested.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "properties": {},
  },
  "stories": [
    {
      "exportName": "Primary",
      "location": [Ln 11, Col 13] - [Ln 11, Col 20],
      "properties": {
        "args": {
          "children": "Nested redundant filename",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/autoTitle/csf/withIndex/index.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "properties": {},
  },
  "stories": [
    {
      "exportName": "Primary",
      "location": [Ln 11, Col 13] - [Ln 11, Col 20],
      "properties": {
        "args": {
          "children": "Nested index filename",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/features/attached/AttachPoint.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "properties": {},
  },
  "stories": [
    {
      "exportName": "Primary",
      "location": [Ln 11, Col 13] - [Ln 11, Col 20],
      "properties": {
        "args": {
          "children": "Primary",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/features/attached/autodoc-override/AttachPoint.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "properties": {},
  },
  "stories": [
    {
      "exportName": "Primary",
      "location": [Ln 11, Col 13] - [Ln 11, Col 20],
      "properties": {
        "args": {
          "children": "Primary",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/BasicCsf.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 11, Col 1],
    "properties": {
      "args": {
        "color": "blue",
      },
      "title": "Example/Basic CSF",
    },
  },
  "stories": [
    {
      "exportName": "Primary",
      "location": [Ln 15, Col 13] - [Ln 15, Col 20],
      "properties": {
        "args": {
          "children": "Primary",
        },
      },
    },
    {
      "exportName": "Secondary",
      "location": [Ln 20, Col 13] - [Ln 20, Col 22],
      "properties": {
        "args": {
          "children": "Secondary",
        },
      },
    },
    {
      "exportName": "Large",
      "location": [Ln 25, Col 13] - [Ln 25, Col 18],
      "properties": {
        "args": {
          "children": "Large",
          "size": 24,
        },
      },
    },
    {
      "exportName": "Small",
      "location": [Ln 31, Col 13] - [Ln 31, Col 18],
      "properties": {
        "args": {
          "children": "Small",
          "size": 12,
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/CaseLower.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 9, Col 15] - [Ln 11, Col 1],
    "properties": {
      "title": "Example/cas_ing/lower",
    },
  },
  "stories": [
    {
      "exportName": "Basic",
      "location": [Ln 4, Col 13] - [Ln 4, Col 18],
      "properties": {
        "args": {
          "children": "Basic",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/CaseUpper.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 9, Col 15] - [Ln 11, Col 1],
    "properties": {
      "title": "Example/CAS-ING/Upper",
    },
  },
  "stories": [
    {
      "exportName": "Basic",
      "location": [Ln 4, Col 13] - [Ln 4, Col 18],
      "properties": {
        "args": {
          "children": "Basic",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/CsfV1.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 1, Col 15] - [Ln 3, Col 1],
    "properties": {
      "title": "Example/CSF v1",
    },
  },
  "stories": [
    {
      "exportName": "V1StoryName",
      "location": [Ln 5, Col 13] - [Ln 5, Col 24],
      "properties": {
        "story": {
          "name": "Renamed via non-hoisted story",
        },
      },
    },
    {
      "exportName": "V2V1StoryName",
      "location": [Ln 11, Col 13] - [Ln 11, Col 26],
      "properties": {
        "story": {
          "name": "v1 name should be overridden",
        },
        "storyName": "v2 name override",
      },
    },
    {
      "exportName": "V2V1StoryNameEmpty",
      "location": [Ln 18, Col 13] - [Ln 18, Col 31],
      "properties": {
        "story": {
          "name": "v1 name should be used",
        },
        "storyName": "",
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/CustomId.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "id": "Manually specified CSF story ID",
    "location": [Ln 1, Col 15] - [Ln 5, Col 1],
    "properties": {
      "id": "Manually specified CSF story ID",
      "tags": [
        "autodocs",
      ],
      "title": "Example/Custom ID CSF",
    },
  },
  "stories": [
    {
      "exportName": "Story",
      "location": [Ln 7, Col 13] - [Ln 7, Col 18],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/CustomName.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 12, Col 1],
    "properties": {
      "args": {
        "children": "Default content",
        "color": "orange",
      },
      "title": "Example/Custom Name",
    },
  },
  "stories": [
    {
      "exportName": "CustomName",
      "location": [Ln 16, Col 13] - [Ln 16, Col 23],
      "properties": {
        "args": {
          "children": "Expected name: Custom story name",
        },
        "storyName": "Custom story name",
      },
    },
    {
      "exportName": "Overridden",
      "location": [Ln 22, Col 13] - [Ln 22, Col 23],
      "properties": {
        "args": {
          "children": "Expected name: Custom story name override",
        },
        "storyName": "Custom story name override",
      },
    },
    {
      "exportName": "NamedViaVariable",
      "location": [Ln 29, Col 13] - [Ln 29, Col 29],
      "properties": {
        "args": {
          "children": "Expected name: Custom story name var",
        },
        "storyName": "Custom story name var",
      },
    },
    {
      "exportName": "InterpolatedName",
      "location": [Ln 36, Col 13] - [Ln 36, Col 29],
      "properties": {
        "storyName": "Custom story name var in template string",
      },
    },
    {
      "exportName": "EmptyName",
      "location": [Ln 39, Col 13] - [Ln 39, Col 22],
      "properties": {
        "storyName": "",
      },
    },
    {
      "exportName": "LeadingWhitespace",
      "location": [Ln 42, Col 13] - [Ln 42, Col 30],
      "properties": {
        "storyName": "     Leading whitespace",
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/Decorators.stories.tsx 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 3, Col 15] - [Ln 5, Col 1],
    "properties": {
      "title": "Example/Decorators",
    },
  },
  "stories": [
    {
      "exportName": "DecoratedClass",
      "location": [Ln 32, Col 13] - [Ln 32, Col 27],
      "properties": {},
    },
    {
      "exportName": "DecoratedClassWithArgs",
      "location": [Ln 34, Col 13] - [Ln 34, Col 35],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/DeepTree.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 3, Col 15] - [Ln 6, Col 1],
    "properties": {
      "title": "Example/Deep/Tree/With/Several/Kinds",
    },
  },
  "stories": [
    {
      "exportName": "Story",
      "location": [Ln 10, Col 13] - [Ln 10, Col 18],
      "properties": {
        "args": {
          "children": "Deep hierarchy",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/Excluded.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 4, Col 15] - [Ln 7, Col 1],
    "properties": {
      "excludeStories": [
        "excludedExport",
      ],
      "title": "Test/Exclusions/Excluded",
    },
  },
  "stories": [
    {
      "exportName": "excludedExport",
      "location": [Ln 1, Col 13] - [Ln 1, Col 27],
      "properties": {},
    },
    {
      "exportName": "otherExport",
      "location": [Ln 2, Col 13] - [Ln 2, Col 24],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/ExportStyles.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 11, Col 1],
    "properties": {
      "args": {
        "color": "green",
      },
      "title": "Example/Export Styles",
    },
  },
  "stories": [
    {
      "exportName": "NormalExport",
      "location": [Ln 15, Col 6] - [Ln 15, Col 18],
      "properties": {
        "args": {
          "children": "Export Style/Normal",
        },
      },
    },
    {
      "exportName": "AlternativeExportName",
      "location": [Ln 20, Col 6] - [Ln 20, Col 16],
      "properties": {
        "args": {
          "children": "Export Style/Exported as",
        },
      },
    },
    {
      "exportName": "ExportNameOverridden",
      "location": [Ln 25, Col 6] - [Ln 25, Col 29],
      "properties": {
        "args": {
          "children": "Export Style/Story name takes priority over modified export name",
        },
        "storyName": "Override name 3",
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/Form.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 3, Col 15] - [Ln 5, Col 1],
    "properties": {
      "title": "Example/Form",
    },
  },
  "stories": [
    {
      "exportName": "Form",
      "location": [Ln 7, Col 13] - [Ln 7, Col 17],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/Hoisting.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "properties": {
      "title": "Example/Hoisting",
    },
  },
  "stories": [
    {
      "exportName": "Hoisting",
      "location": [Ln 3, Col 13] - [Ln 3, Col 21],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/Included.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 4, Col 15] - [Ln 7, Col 1],
    "properties": {
      "includeStories": [
        "includedExport",
      ],
      "title": "Test/Exclusions/Included",
    },
  },
  "stories": [
    {
      "exportName": "includedExport",
      "location": [Ln 1, Col 13] - [Ln 1, Col 27],
      "properties": {},
    },
    {
      "exportName": "otherExport",
      "location": [Ln 2, Col 13] - [Ln 2, Col 24],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/JsxExtension.stories.jsx 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 1, Col 15] - [Ln 3, Col 1],
    "properties": {
      "title": "Example/JSX extension",
    },
  },
  "stories": [
    {
      "exportName": "Story",
      "location": [Ln 5, Col 13] - [Ln 5, Col 18],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/MergeOne.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 1, Col 15] - [Ln 3, Col 1],
    "properties": {
      "title": "Example/Merge",
    },
  },
  "stories": [
    {
      "exportName": "One",
      "location": [Ln 5, Col 13] - [Ln 5, Col 16],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/MergeTwo.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 1, Col 15] - [Ln 3, Col 1],
    "properties": {
      "title": "Example/Merge",
    },
  },
  "stories": [
    {
      "exportName": "Two",
      "location": [Ln 5, Col 13] - [Ln 5, Col 16],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/SameName.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 8, Col 1],
    "properties": {
      "title": "Example/With Same Name",
    },
  },
  "stories": [
    {
      "exportName": "SameName",
      "location": [Ln 12, Col 13] - [Ln 12, Col 21],
      "properties": {
        "args": {
          "children": "Same name story",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/SameNameChild.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 8, Col 1],
    "properties": {
      "title": "Example/With Same Name/Same Name",
    },
  },
  "stories": [
    {
      "exportName": "NestedStory",
      "location": [Ln 12, Col 13] - [Ln 12, Col 24],
      "properties": {
        "args": {
          "children": "Nested same name story",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/StoryDeclarations.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 1, Col 6] - [Ln 3, Col 1],
    "properties": {
      "title": "Test/Story Declarations",
    },
  },
  "stories": [
    {
      "exportName": "ConstDeclarator",
      "location": [Ln 7, Col 13] - [Ln 7, Col 28],
      "properties": {},
    },
    {
      "exportName": "SeparateDeclarator",
      "location": [Ln 9, Col 6] - [Ln 9, Col 24],
      "properties": {},
    },
    {
      "exportName": "SeparateDeclarator2",
      "location": [Ln 10, Col 6] - [Ln 10, Col 25],
      "properties": {
        "storyName": "Separate declarator with story name",
      },
    },
    {
      "exportName": "MultipleOne",
      "location": [Ln 15, Col 13] - [Ln 15, Col 24],
      "properties": {},
    },
    {
      "exportName": "MultipleTwo",
      "location": [Ln 16, Col 2] - [Ln 16, Col 13],
      "properties": {
        "storyName": "Multiple two (given name)",
      },
    },
    {
      "exportName": "FunctionStory",
      "location": [Ln 20, Col 16] - [Ln 20, Col 29],
      "properties": {
        "storyName": "Renamed Function Story",
      },
    },
    {
      "exportName": "SeparateFunctionStory",
      "location": [Ln 25, Col 9] - [Ln 25, Col 30],
      "properties": {},
    },
    {
      "exportName": "SeparateFunctionStory2",
      "location": [Ln 29, Col 9] - [Ln 29, Col 31],
      "properties": {
        "storyName": "Separate function story with story name",
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/TaggedDocs.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 12, Col 1],
    "properties": {
      "args": {
        "color": "blue",
      },
      "tags": [
        "autodocs",
      ],
      "title": "Example/Tagged autodocs",
    },
  },
  "stories": [
    {
      "exportName": "Primary",
      "location": [Ln 16, Col 13] - [Ln 16, Col 20],
      "properties": {
        "args": {
          "children": "Primary",
        },
      },
    },
    {
      "exportName": "Secondary",
      "location": [Ln 21, Col 13] - [Ln 21, Col 22],
      "properties": {
        "args": {
          "children": "Secondary",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/Top.stories.js 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 8, Col 1],
    "properties": {
      "title": "Top",
    },
  },
  "stories": [
    {
      "exportName": "TopLevelStory",
      "location": [Ln 12, Col 13] - [Ln 12, Col 26],
      "properties": {
        "args": {
          "children": "Top level story",
        },
      },
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/TypeScript.stories.tsx 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 17, Col 15] - [Ln 19, Col 9],
    "properties": {
      "title": "Test/TypeScript/Sample",
    },
  },
  "stories": [
    {
      "exportName": "TypeScriptStory",
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "properties": {},
    },
    {
      "exportName": "TypeScriptTypedStory",
      "location": [Ln 5, Col 13] - [Ln 5, Col 40],
      "properties": {},
    },
    {
      "exportName": "TypeScriptTypedGenericStory",
      "location": [Ln 9, Col 13] - [Ln 9, Col 51],
      "properties": {},
    },
    {
      "exportName": "TypeScriptTypeAssertionStory",
      "location": [Ln 13, Col 13] - [Ln 13, Col 41],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/TypeScriptAsMeta.stories.ts 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 15] - [Ln 7, Col 9],
    "properties": {
      "title": "Test/TypeScript/Default as Meta",
    },
  },
  "stories": [
    {
      "exportName": "TypeScriptStory",
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/TypeScriptIdentifier.stories.ts 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 5, Col 6] - [Ln 7, Col 1],
    "properties": {
      "title": "Test/TypeScript/Typed Identifier",
    },
  },
  "stories": [
    {
      "exportName": "TypeScriptStory",
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/TypeScriptIdentifierAsMeta.stories.ts 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 6, Col 6] - [Ln 8, Col 9],
    "properties": {
      "title": "Test/TypeScript/Identifier as Meta",
    },
  },
  "stories": [
    {
      "exportName": "TypeScriptStory",
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "properties": {},
    },
  ],
}
`;

exports[`csf > parses story file project/v7/src/stories/TypeScriptIdentifierSatisfies.stories.ts 1`] = `
{
  "meta": {
    "declared": true,
    "location": [Ln 6, Col 6] - [Ln 8, Col 16],
    "properties": {
      "title": "Test/TypeScript/Identifier Satisfies",
    },
  },
  "stories": [
    {
      "exportName": "TypeScriptStory",
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "properties": {},
    },
  ],
}
`;
